cpp/src/arrow/meson.build

```diff
@@ -0,0 +1,333 @@
+        'dependencies': [],
+    },
+    'memory_pool': {
+        'sources': ['memory_pool.cc'],
+        'include_dirs': [include_dir],
+        'dependencies': [],
+    },
+    'vendored': {
+        'sources': [
+            'vendored/base64.cpp',
+            'vendored/datetime.cpp',
+            'vendored/double-conversion/bignum-dtoa.cc',
+            'vendored/double-conversion/bignum.cc',
+            'vendored/double-conversion/cached-powers.cc',
+            'vendored/double-conversion/double-to-string.cc',
+            'vendored/double-conversion/fast-dtoa.cc',
+            'vendored/double-conversion/fixed-dtoa.cc',
+            'vendored/double-conversion/string-to-double.cc',
+            'vendored/double-conversion/strtod.cc',
+            'vendored/musl/strptime.c',
+            'vendored/uriparser/UriCommon.c',
+            'vendored/uriparser/UriCompare.c',
+            'vendored/uriparser/UriEscape.c',
+            'vendored/uriparser/UriFile.c',
+            'vendored/uriparser/UriIp4.c',
+            'vendored/uriparser/UriIp4Base.c',
+            'vendored/uriparser/UriMemory.c',
+            'vendored/uriparser/UriNormalize.c',
+            'vendored/uriparser/UriNormalizeBase.c',
+            'vendored/uriparser/UriParse.c',
+            'vendored/uriparser/UriParseBase.c',
+            'vendored/uriparser/UriQuery.c',
+            'vendored/uriparser/UriRecompose.c',
+            'vendored/uriparser/UriResolve.c',
+            'vendored/uriparser/UriShorten.c',
+        ],
+        'include_dirs': [include_dir],
+        'dependencies': [],
+    },
+}
+
+objlibs = []
+foreach key, val : objlib_sources
+    slib = static_library(
+        key,
+        sources: val['sources'],
+        include_directories: val['include_dirs'],
+        dependencies: val['dependencies'],
+        build_by_default: false,
+    )
+    objlibs += slib.extract_all_objects(recursive: true)
+endforeach
+
+arrow_srcs = [
+    'builder.cc',
+    'buffer.cc',
+    'chunked_array.cc',
+    'chunk_resolver.cc',
+    'compare.cc',
+    'config.cc',
+    'datum.cc',
+    'device.cc',
+    'device_allocation_type_set.cc',
+    'extension_type.cc',
+    'extension/bool8.cc',
+    'extension/json.cc',
+    'extension/uuid.cc',
+    'pretty_print.cc',
+    'record_batch.cc',
+    'result.cc',
+    'scalar.cc',
+    'sparse_tensor.cc',
+    'status.cc',
+    'table.cc',
+    'table_builder.cc',
+    'tensor.cc',
+    'tensor/coo_converter.cc',
+    'tensor/csf_converter.cc',
+    'tensor/csx_converter.cc',
+    'type.cc',
+    'type_traits.cc',
+    'visitor.cc',
+    'c/bridge.cc',
+    'c/dlpack.cc',
+]
+
+arrow_lib = library(
+    'arrow',
+    sources: [arrow_srcs],
+    objects: objlibs,
+    include_directories: [include_dir],
+    install: true,
+    # compute/expression.cc may have undefined IPC symbols in non-IPC builds
+    override_options: {'b_lundef': 'false'},
+)
+
+# Meson does not allow you to glob for headers to install. See also
```

Comment:
Meson is pretty strict about wildcards for the reasons outlined in their FAQ. Idiomatically, Meson would want you to put the files you want to install in a separate directory and call `install_subdir`, but that would go beyond the scope of this initial PR I think